lib/active_triples/entity.rb

```diff
@@ -0,0 +1,551 @@
+
+    attr_accessor :parent
+
+    def type_registry
+      @@type_registry ||= {}
+    end
+    module_function :type_registry
+
+    included do
+      extend Configurable
+      extend Deprecation
+      extend ActiveModel::Naming
+      extend ActiveModel::Translation
+      extend ActiveModel::Callbacks
+
+      delegate :each, :load!, :count, :has_statement?, :to => :@graph
+
+      define_model_callbacks :persist
+
+      ##
+      # Adapter for a consistent interface for creating a new Resource
+      # from a URI. Similar functionality should exist in all objects
+      # which can become a Resource.
+      #
+      # @param uri [#to_uri, String]
+      # @param vals values to pass as arguments to ::new
+      #
+      # @return [ActiveTriples::Entity] a Resource with the given uri
+      def self.from_uri(uri, vals = nil)
+        new(uri, vals)
+      end
+
+      ##
+      # Test if the rdf_subject that would be generated using a
+      # specific ID is already in use in the triplestore.
+      #
+      # @param [Integer, #read] ID to test
+      #
+      # @return [TrueClass, FalseClass] true, if the ID is in
+      #    use in the triplestore; otherwise, false.
+      #    NOTE: If the ID is in use in an object not yet
+      #          persisted, false will be returned presenting
+      #          a window of opportunity for an ID clash.
+      def self.id_persisted?(test_id)
+        rdf_subject = self.new(test_id).rdf_subject
+        ActiveTriples::Repositories.has_subject?(rdf_subject)
+      end
+
+
+      ##
+      # Test if the rdf_subject that would be generated using a
+      # specific URI is already in use in the triplestore.
+      #
+      # @param [String, RDF::URI, #read] URI to test
+      #
+      # @return [TrueClass, FalseClass] true, if the URI is in
+      #    use in the triplestore; otherwise, false.
+      #    NOTE: If the URI is in use in an object not yet
+      #          persisted, false will be returned presenting
+      #          a window of opportunity for an ID clash.
+      def self.uri_persisted?(test_uri)
+        rdf_subject = test_uri.kind_of?(RDF::URI) ? test_uri : RDF::URI(test_uri)
+        ActiveTriples::Repositories.has_subject?(rdf_subject)
+      end
+
+      protected
+
+      def insert_statement(*args)
+        @graph.send(:insert_statement, *args)
+      end
+
+      def delete_statement(*args)
+        @graph.send(:delete_statement, *args)
+      end
+    end
+
+    ##
+    # Specifies whether the object is currently writable.
+    #
+    # @return [true, false]
+    def writable?
+      !frozen?
+    end
+
+    ##
+    # Initialize an instance of this resource class. Defaults to a
+    # blank node subject. In addition to RDF::Graph parameters, you
+    # can pass in a URI and/or a parent to build a resource from a
+    # existing data.
+    #
+    # You can pass in only a parent with:
+    #    new(nil, parent)
+    #
+    # @see RDF::Graph
+    def initialize(*args, &block)
```

Comment:
Fixing this could be a pretty big refactor. In my mind the way forward is likely to allow a protected setter for #rdf_subject and let the implementor take care of setting that and the persistence point.

Then there's that type business. We should probably find a better way to do that anyways.